### PR TITLE
feat: adds `live` function to create dynamic console applications

### PR DESCRIPTION
**Pull request by nunomaduro**
This pull request allows to make a particular section of the console dynamic. Here is an example:

```php
use function Termwind\{live};

live(function () {
    static $counter = 0;

    $counter++;

    return '<div class="m-1">
        <span class="pl-1 pr-2 bg-blue-300">My counter</span>
        <span class="ml-1">' . $counter . '</span>
    </div>';
})->refreshEvery(seconds: 1);
```

Will render the following:
live

In addition, you may control when the refresh should stop, by calling the `RefreshEvent::stop` method.

```php
live(function (RefreshEvent $event) {
    return $event->stop();
})->refreshEvery(seconds: 1);
```

Finally, if you wish to control, when to render, clear, or refresh, you may control the flow using these methods:

```php
use function Termwind\{live};

$live = live(fn () => 'html');

$live->render(); // Re-runs the given closure, and renders its output.
$live->clear(); // Removes all previous rendered html.
$live->refresh(); // Refreshes the previous rendered html.
$live->refreshEvery($seconds); // Refreshes the previous rendered html, every X amount of seconds.
```

### CODE REVIEW

**Review by xiCO2k — APPROVED** [review]

File: src/Functions.php [hunk 1]
```diff
@@ -51,3 +53,25 @@ function terminal(): Terminal
         return new Terminal;
     }
 }
+
+if (! function_exists('live')) {
```
Comment (mabar) [inline src/Functions.php:57]: I believe this is not neccessary because the function is in namespace? Also it should be `function_exists('\Termwind\live')`

Comment (nunomaduro) [reply src/Functions.php:57]: Can you pull request a fix please?

**Comment by mabar** [comment]
Do you plan to also add ability to react on input or is there some other way? I would like to use this API for some console games :)

File: src/Functions.php (hunk 1 above)
Comment (mabar) [reply src/Functions.php:57]: @nunomaduro Sure. Do you prefer to add namespace to check or remove the check completely?

Comment (nunomaduro) [reply src/Functions.php:57]: Add the namespace check.

**Comment by nunomaduro** [comment]
@mabar sure - can you discuss this with @xiCO2k

**Comment by mabar** [comment]
@xiCO2k I don't really know how it works in php, but enabling refresh on input and having ability to get input key like in javascript could be enough. JS codes: https://keycode.info

```php
use function Termwindlive;

live(function (RefreshReason $reason) {
    if ($reason->isOnInputRefresh() && $reason->getInputKeyCode() === 'ArrowUp') {
    	// ...
    }

	if ($reason->isScheduledRefresh()) {
		// ...
	}


	if ($reason->isManualRefresh()) {
		// ...
	}

    return '';
})->refreshEvery(seconds: 1)->refreshOnInput();
```

Only thing that I am worried about is the user holding key and whether it's gonna spam the event every few miliseconds

**Comment by xiCO2k** [comment]
This could be pretty interesting, will definitely need to check if there is a way to have handle those type of events.

**Comment by xiCO2k** [comment]
ping @nunomaduro.

**Comment by nunomaduro** [comment]
I will need to look into this pr seriously. Sorry. Just in January.

**Comment by xiCO2k** [comment]
Got it just to make sure it was not missed 💪

**Comment by mabar** [comment]
I found this neat little [snake game](https://github.com/dbu/snake-bundle) which may have a way to go:

It's an infinite loop which simulates ticks via measuring time of each cycle and sleeps in case any time of tick is left
https://github.com/dbu/snake-bundle/blob/e31451f87bc63c497e5defdd3eac5a2690295108/src/Game/Game.php#L54-L83

Only thing that seems "extra" is setting a non-blocking stream and executing `stty` in shell
https://github.com/dbu/snake-bundle/blob/e31451f87bc63c497e5defdd3eac5a2690295108/src/Command/SnakeCommand.php#L30-L47

For tracking movement is used [Cursor](https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/Console/Cursor.php) from Symfony and [Coordinate](https://github.com/dbu/snake-bundle/blob/e31451f87bc63c497e5defdd3eac5a2690295108/src/Game/Coordinate.php) value object.

**Comment by xiCO2k** [comment]
> I found this neat little [snake game](https://github.com/dbu/snake-bundle) which may have a way to go:
> 
> It's an infinite loop which simulates ticks via measuring time of each cycle and sleeps in case any time of tick is left https://github.com/dbu/snake-bundle/blob/e31451f87bc63c497e5defdd3eac5a2690295108/src/Game/Game.php#L54-L83
> 
> Only thing that seems "extra" is setting a non-blocking stream and executing `stty` in shell https://github.com/dbu/snake-bundle/blob/e31451f87bc63c497e5defdd3eac5a2690295108/src/Command/SnakeCommand.php#L30-L47
> 
> For tracking movement is used [Cursor](https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/Console/Cursor.php) from Symfony and [Coordinate](https://github.com/dbu/snake-bundle/blob/e31451f87bc63c497e5defdd3eac5a2690295108/src/Game/Coordinate.php) value object.

Thanks for that, I'm playing around with it on a new feature that will be added after this one, that will allow to have cursor events.

More info here: https://twitter.com/xiCO2k/status/1473348536821751808

**Comment by nunomaduro** [comment]
@xiCO2k As discussed, can you close this issue, and maybe make a separate project for this?

**Comment by xiCO2k** [comment]
Yes will do that, thanks @nunomaduro

**Pull request closed without merging**